File: neetcode all and misc questions/Stacks and Queues/baseball_game.py
"""
https://leetcode.com/problems/baseball-game/
"""
class Solution:
    """ Revisited 10/8/2024 """
    def calPoints(self, operations: List[str]) -> int:
        stack = []
        for i in range(len(operations)):
            if operations[i] == "+" and len(stack) > 1:
                stack.append(stack[-1] + stack[-2])
            elif operations[i] == "D" and len(stack) > 0:
                stack.append(stack[-1] * 2)
            elif operations[i] == "C" and len(stack) > 0:
                stack.pop()
            else:
                stack.append(int(operations[i]))
        return sum(stack)

class Solution:
    def calPoints(self, operations: List[str]) -> int:
        stack = []
        for i in range(len(operations)):
            isInt = False
            try:
                num = int(operations[i])
                isInt = True
            except:
                isInt = False
            if isInt:
                stack.append(int(operations[i]))
            elif operations[i] == "C":
                stack.pop()
            elif operations[i] == "D":
                stack.append(stack[-1] * 2)
            elif operations[i] == "+":
                stack.append(stack[-1] + stack[-2])
        return sum(stack)

File: node_js_ds_and_algos_implementations/stacks/baseball_game.js
class Solution {
    /**
     * @param {string[]} operations
     * @return {number}
     */
    calPoints(operations) {
        /* 
        stack problem
        [1,2,+]
        look at the last two, add them together and push (without popping out)
        [1, 2, 3, "C"]
        [1, 2], invalidate the previous score (pop out 3)
        [1, 2, 5, C], double the previous score => 10
        [1,2,5,10]
        sum together to get 18
        */
        let stack = []
        for (let i = 0; i < operations.length; ++i){
            if (stack.length){
                // there should always be at least two numbers on the stack before "+"
                if (operations[i] === "+"){
                    stack.push(stack[stack.length-1] + stack[stack.length-2])
                }
                else if (operations[i] === "C"){
                    stack.pop()
                }
                else if (operations[i] === "D"){
                    stack.push(stack[stack.length-1] * 2)
                } 
                else {
                    stack.push(Number(operations[i]))
                }
                continue
            }
            else {
                stack.push(Number(operations[i]))
            }
        }
        return stack.reduce((acc, total) => acc + total, 0)

    }
